src/validation/rules/UniqueArgumentDefinitionNamesRule.ts

```diff
@@ -0,0 +1,91 @@
+import { GraphQLError } from '../../error/GraphQLError';
+
+import type { ASTVisitor } from '../../language/visitor';
+import type {
+  NameNode,
+  FieldDefinitionNode,
+  InputValueDefinitionNode,
+} from '../../language/ast';
+
+import type { SDLValidationContext } from '../ValidationContext';
+
+/**
+ * Unique argument definition names
+ *
+ * A GraphQL Object or Interface type is only valid if all its fields have uniquely named arguments.
+ * A GraphQL Directive is only valid if all its arguments are uniquely named.
+ */
+export function UniqueArgumentDefinitionNamesRule(
+  context: SDLValidationContext,
+): ASTVisitor {
+  return {
+    DirectiveDefinition(directiveNode) {
+      // istanbul ignore next (See: 'https://github.com/graphql/graphql-js/issues/2203')
+      const argumentNodes = directiveNode.arguments ?? [];
+
+      return checkArgUniqueness(`@${directiveNode.name.value}`, argumentNodes);
+    },
+    InterfaceTypeDefinition: checkArgUniquenessPerField,
+    InterfaceTypeExtension: checkArgUniquenessPerField,
+    ObjectTypeDefinition: checkArgUniquenessPerField,
+    ObjectTypeExtension: checkArgUniquenessPerField,
+  };
+
+  function checkArgUniquenessPerField(typeNode: {
+    readonly name: NameNode;
+    readonly fields?: ReadonlyArray<FieldDefinitionNode>;
+  }) {
+    const typeName = typeNode.name.value;
+
+    // istanbul ignore next (See: 'https://github.com/graphql/graphql-js/issues/2203')
+    const fieldNodes = typeNode.fields ?? [];
+
+    for (const fieldDef of fieldNodes) {
```

Comment:
Why split this out into its own, used-once const, rather than:
```suggestion
    for (const fieldDef of typeNode.fields ?? []) {
```